cpp/src/arrow/dataset/file_base.cc

```diff
@@ -89,6 +89,12 @@ Result<std::shared_ptr<io::InputStream>> FileSource::OpenCompressed(
   return io::CompressedInputStream::Make(codec.get(), std::move(file));
 }
 
+bool FileSource::Equals(const FileSource& other) const {
+  if (filesystem_ == NULLPTR) return false;
```

Comment:
This shouldn't be possible right?

Also `NULLPTR` is meant for headers, not source files

Reply:
Mmm sorry didn’t catch this one clearly. Could you explain a bit?

Reply:
Can filesystem_ be null? I would have assumed not given this is a FileFragment (and it should be validated in the constructor)